zio-schema/shared/src/main/scala/zio/schema/meta/ExtensibleMetaSchema.scala

```diff
@@ -0,0 +1,724 @@
+package zio.schema.meta
+
+import scala.annotation.tailrec
+import scala.collection.immutable.ListMap
+import scala.collection.mutable
+
+import zio.constraintless.TypeList
+import zio.prelude._
+import zio.schema._
+import zio.{ Chunk, ChunkBuilder }
+
+sealed trait ExtensibleMetaSchema[BuiltIn <: TypeList] { self =>
+  def builtInInstances: SchemaInstances[BuiltIn]
+
+  def path: NodePath
+  def optional: Boolean
+
+  def toSchema: Schema[_] = {
+    val refMap = mutable.HashMap.empty[NodePath, Schema[_]]
+    ExtensibleMetaSchema.materialize(self, refMap)(builtInInstances)
+  }
+
+  override def toString: String = AstRenderer.render(self)
+}
+
+object ExtensibleMetaSchema {
+  import CaseSet._
+
+  type Labelled[BuiltIn <: TypeList] = (String, ExtensibleMetaSchema[BuiltIn])
+  type Lineage                       = Chunk[(Int, NodePath)]
```

Comment:
If these two are exposed to the outside world, may be better to make them `case class`.

Reply:
I agree. They were type aliases in the existing `MetaSchema` and I did not want to change more than necessary. Should I do it in this PR?

Reply:
Changed